### PR TITLE
Support sdk-type binding reference type

### PR DESCRIPTION
**Pull request by liliankasem**
### Issue describing the changes in this PR

#1081

### Pull request checklist

* [ ] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs
Comment (mathewc) [inline extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs:17]: Wonder if this shouldn't be named SupportsWorkerParameterBinding?

File: extensions/Worker.Extensions.Abstractions/src/ReferenceTypeAttribute.cs
Comment (mathewc) [inline extensions/Worker.Extensions.Abstractions/src/ReferenceTypeAttribute.cs:9]: Once we agree on the naming above, this attribute would follow suit. E.g. it could be called SupportsWorkerParameterBinding. And likely it should just be a marker attribute, not bool param needed?

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (mathewc) [inline sdk/Sdk/FunctionMetadataGenerator.cs:429]: nit - adjust comment to "For extensions that support worker binding, set the DataType so all parameters are bound by the worker."

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (mathewc) [inline sdk/Sdk/FunctionMetadataGenerator.cs:705]: nit: remove unnecessary comment?

File: src/DotNetWorker.Core/IBindingData.cs
Comment (mathewc) [inline src/DotNetWorker.Core/IBindingData.cs:11]: do we really need this interface? Doesn't seem like other types are following this pattern.

File: extensions/Worker.Extensions.Abstractions/src/ReferenceTypeAttribute.cs
Comment (liliankasem) [reply extensions/Worker.Extensions.Abstractions/src/ReferenceTypeAttribute.cs:9]: Ah this should be deleted.

We had discussed either introducing a new attribute (like this one) or using the existing extension information attribute to determine if an extension supports reference types. In the end we went with using the existing attribute so I should have removed this file.

In the `ExtensionInformationAttribute` file, we introduced a `SupportsBindingReferenceType` property - do you feel we should rename this and not use bool for that?

File: src/DotNetWorker.Core/IBindingData.cs
Comment (liliankasem) [reply src/DotNetWorker.Core/IBindingData.cs:11]: We use this interface in the converters to determine if the converter can handle the request or not:

https://github.com/Azure/azure-functions-dotnet-worker/blob/92c5be67bcda0c3badd980ec2524f7c43eb7d859/extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs#L23-L26

File: src/DotNetWorker.Grpc/BindingData.cs
Comment (mathewc) [inline src/DotNetWorker.Grpc/BindingData.cs:9]: For the naming of the types that are consumed by worker converters, I wonder if ParameterBindingData wouldn't be better here?

**Comment by surgupta-msft** [comment]
ToDo - Adding tests

File: src/DotNetWorker.Grpc/BindingData.cs
Comment (liliankasem) [reply src/DotNetWorker.Grpc/BindingData.cs:9]: Renaming to ModelBindingData

File: src/DotNetWorker.Grpc/protobuf/src/proto/FunctionRpc.proto
Comment (liliankasem) [inline src/DotNetWorker.Grpc/protobuf/src/proto/FunctionRpc.proto:32]: Make sure to remove proto changes in this PR once we have a new PR to bring in the new proto version

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (liliankasem) [inline sdk/Sdk/FunctionMetadataGenerator.cs:468]: We probably don't need to have a check for output binding here because when we call this method (`BuildBindingMetadataFromAttribute`) from the output binding path, we always set supportsReferenceType to false

https://github.com/Azure/azure-functions-dotnet-worker/blob/08e86a23c00fd8199639c4588397b88c6bebdcae/sdk/Sdk/FunctionMetadataGenerator.cs#L444-L449

File: extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs
Comment (liliankasem) [reply extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs:17]: With BindingData renamed to ModelBindingData, would `SupportsWorkerModelBinding` make sense?

Comment (mathewc) [reply extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs:17]: If we go with the defer bind naming, then this could be SupportsDeferredBinding

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (mathewc) [inline sdk/Sdk/FunctionMetadataGenerator.cs:384]: We need defer bind if any input OR output bindings have binding expressions, right? For example, otherwise, if we were to defer bind for the trigger, the host may not have the binding metadata it needs to resolve the output binding, resulting in a runtime error.

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (mathewc) [inline sdk/Sdk/FunctionMetadataGenerator.cs:401]: You're missing a check here IsFunctionBindingType which is done in the other loop above. Why do we need to loop through all the parameters twice like this anyways? You could just loop through once, build a list of binding metadatas, then at the end set the deferred property based on whether any used expressions.

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (liliankasem) [reply sdk/Sdk/FunctionMetadataGenerator.cs:384]: That's a good point. I think this class in general is confusing as this current function is called "AddInputTriggerBindingsAndExtensions" and so at this point I assume we only have input and triggers, but there's nothing upstream making that split between input/trigeer and output. I might need to get some clarity from @fabiocav and @brettsam about this, but in general agree with your point about triggers deferring binding if outputs also contain expressions

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (liliankasem) [reply sdk/Sdk/FunctionMetadataGenerator.cs:401]: This was definitly a brute force attempt to get a prototype running so definitly not expecting this to be final form :D I think your suggestion makes sense, you're saying that at the end we can just loop through `bindingMetadata` to check if any input/output bindings use expressions and use that to change the flag for the trigger? This should be a lot more efficient

File: src/DotNetWorker.Core/IBindingData.cs
Comment (liliankasem) [reply src/DotNetWorker.Core/IBindingData.cs]: Hmm, that's fair. I think the intension was around project dependencies. ModelBindingData is defined in the Grpc project, and the extension projects will be [updated to reference the Core project](https://github.com/Azure/azure-functions-dotnet-worker/blob/568045c48fcda451a62b691f34c948ad3d429dbc/extensions/Worker.Extensions.Storage.Blobs/src/Worker.Extensions.Storage.Blobs.csproj#L19), this is so they have access to converter interfaces such as `IInputConverter`. And I believe we didn't want the extensions to take a dependency on the Grpc layer. If that isn't a good enough reason, then I am happy to modify this and work without an interface.

@fabiocav want to get your input here too as you helped make the initial interface decision

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (surgupta-msft) [reply sdk/Sdk/FunctionMetadataGenerator.cs]: Done

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (surgupta-msft) [reply sdk/Sdk/FunctionMetadataGenerator.cs]: Done

File: extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs
Comment (liliankasem) [inline extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs:17]: @jviau FYI: I think you commented here about this being `public bool SupportsDeferredBinding { get; init; }` but in the blob converter PR, the converter PR wasn't meant to have these base changes so that's my bad on PR setup, but should be fixed now. Just letting you know in case you wanted to do your review here too!

File: pack-sdk.ps1
Comment (mathewc) [inline pack-sdk.ps1:1]: Remove from PR

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (mathewc) [inline sdk/Sdk/FunctionMetadataGenerator.cs:779]: Why are we enabling things based on ctor definitions, rather than an actual instance of the assembly attribute? One problem I see with this is that if the extension specifies false for these, e.g. `[assembly: ExtensionInformation("Microsoft.Azure.WebJobs.Extensions.Storage.Blobs", "5.2.0-alpha.20221028.1", false, false)]` you'll still be enabling. Another problem is that you're doing a strict equality check for 4. If someone adds a 5th ctor overload in the future, it'll break your code.

File: src/DotNetWorker.Core/IModelBindingData.cs
Comment (mathewc) [inline src/DotNetWorker.Core/IModelBindingData.cs:11]: I still don't understand why this needs to be an interface. It seems unnecessary, and also means that we can't add things to the type in the future (e.g. new binding properties) without breaking contract changes.

Comment (liliankasem) [reply src/DotNetWorker.Core/IModelBindingData.cs:11]: I think we have a thread about this here: https://github.com/Azure/azure-functions-dotnet-worker/pull/1107#discussion_r1010958793

I'm not opposed to removing the interface but want to double check with Fabio around his though process - I can sync with him offline and update this

File: src/ci.yml
Comment (surgupta-msft) [inline src/ci.yml:56]: This change will be removed before merging PR. It's only for testing to unblock pipeline.

File: pack-sdk.ps1
Comment (surgupta-msft) [reply pack-sdk.ps1]: Removed

File: src/DotNetWorker.Core/IBindingData.cs
Comment (fabiocav) [reply src/DotNetWorker.Core/IBindingData.cs]: Core types (`Worker.Core`) do not have a dependency on any types coming from the `Worker.Grpc` project, as that is one (of possibly many) transport implementations. Any time the gRPC implementation needs to pass its types to the infrastructure/core, it does so using the abstractions defined there. `IBindingData` is just that, the abstraction that is exposed in the public APIs in Core, which is what is exposed to converters and other components. Those components do not have a dependency on any other transport specific types.

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (surgupta-msft) [reply sdk/Sdk/FunctionMetadataGenerator.cs]: For the last point regarding strict equality check, I have updated the check to >=4. And if Extension specifics false then below code will assign supportsDeferredBinding to False

`supportsDeferredBinding = (bool)assemblyAttribute.ConstructorArguments[3].Value;`

Regarding first question, _Why are we enabling things based on ctor definitions, rather than an actual instance of the assembly attribute?_

Need to check with @fabiocav. I remember an old discussion to extend the logic of evaluating constructor arguments.

File: src/DotNetWorker.Core/IModelBindingData.cs
Comment (surgupta-msft) [reply src/DotNetWorker.Core/IModelBindingData.cs:11]: cc @fabiocav

Comment (surgupta-msft) [reply src/DotNetWorker.Core/IModelBindingData.cs:11]: Resolving this one as we are discussing this under comment - https://github.com/Azure/azure-functions-dotnet-worker/pull/1107#discussion_r1002146117

File: src/DotNetWorker.Core/IBindingData.cs
Comment (mathewc) [reply src/DotNetWorker.Core/IBindingData.cs]: Guess I'm not clear on the layering here then. BindingData could itself probably just be in Worker.Core then? If you retain the interface, and we decide we need to add another bit of info later, is it easy for us to rev both the interface and the impl together?

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (mathewc) [reply sdk/Sdk/FunctionMetadataGenerator.cs]: Once you have the attribute instance, why can't you just check attr.SupportsDeferredBinding? Why are you checking ctor params for any of these things that the attribute exposes properties for?

File: src/DotNetWorker.Core/IBindingData.cs
Comment (fabiocav) [reply src/DotNetWorker.Core/IBindingData.cs]: `BindingData` is an auto-generated type from the `proto` file (it's what is defined in the protobuf contract), internal and also implements some gRPC specific interfaces.

We could use an abstract class instead of the interface, as that would help mitigate the concern. That's what we've done for things like the `HttpRequestData`

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (surgupta-msft) [reply sdk/Sdk/FunctionMetadataGenerator.cs]: I don't think we are getting attribute instance here, instead we are just getting AssemblyDefinition.

```csharp
 private static bool TryAddExtensionInfo(IDictionary<string, string> extensions, AssemblyDefinition extensionAssemblyDefinition, out bool supportsDeferredBinding, bool usedByFunction = true)
        {
            supportsDeferredBinding = false;

            foreach (var assemblyAttribute in extensionAssemblyDefinition.CustomAttributes)
            {
                if (string.Equals(assemblyAttribute.AttributeType.FullName, Constants.ExtensionsInformationType, StringComparison.Ordinal))
                {
                    string extensionName = assemblyAttribute.ConstructorArguments[0].Value.ToString();
                    string extensionVersion = assemblyAttribute.ConstructorArguments[1].Value.ToString();
                    bool implicitlyRegister = false;
                    if (assemblyAttribute.ConstructorArguments.Count >= 3)
                    {
                        // EnableImplicitRegistration
                        implicitlyRegister = (bool)assemblyAttribute.ConstructorArguments[2].Value;
                    }

                    if (assemblyAttribute.ConstructorArguments.Count >= 4)
                    {
                        supportsDeferredBinding = (bool)assemblyAttribute.ConstructorArguments[3].Value;
                    }
```

Comment (fabiocav) [reply sdk/Sdk/FunctionMetadataGenerator.cs]: Yeah, it's a little awkward, but this uses `Mono.Cecil`, that isn't working with an actual instance of the attribute in this case.

File: extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs
Comment (liliankasem) [reply extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs:17]: @surgupta-msft do you want to see if `public bool SupportsDeferredBinding { get; init; }` works, if so then you can remove the new ctor

Here's the original comment for reference: https://github.com/Azure/azure-functions-dotnet-worker/pull/1108#discussion_r1010957806

File: sdk/Sdk/FunctionMetadataGenerator.cs
Comment (mathewc) [reply sdk/Sdk/FunctionMetadataGenerator.cs]: Don't know the backstory on that, but if you're happy, I'm happy

**Review by mathewc — APPROVED** [review]

File: src/DotNetWorker.Core/IBindingData.cs
Comment (surgupta-msft) [reply src/DotNetWorker.Core/IBindingData.cs]: Done

File: extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs
Comment (surgupta-msft) [reply extensions/Worker.Extensions.Abstractions/src/ExtensionInformationAttribute.cs:17]: Since the existing fields of this class are also initialized in similar way, I will not block this PR, but happy to try this refactoring out separately.

**Pull request merged**
